lib/rspec/mocks/proxy.rb

```diff
@@ -371,20 +371,34 @@ def original_method_handle_for(message)
 
         return super unless unbound_method
         unbound_method.bind(object)
+      rescue TypeError
+        if RUBY_VERSION == '1.8.7'
```

Comment:
Instead of checking the ruby version at runtime each time this occurs, we prefer to check the ruby version once at file load time, and conditionally define different implementations of methods based on the version.  For example, you could do something like:

``` ruby
if RUBY_VERSION == '1.8.7'
  def handle_bind_type_error
    yield
  rescue TypeError
    # your logic...
  end
else
  def handle_bind_type_error
    yield
  end
end
```

Then `original_method_handle_for` can wrap the `bind` call in `handle_bind_type_error { ... }` and the appropriate thing will happen, without the cost of checking the ruby version or rescuing and re-raising.

There may be a simpler, better way to organize this, but hopefully that gives you an idea.